### ISSUE TITLE
New Rule: No potential spread overwrites

### ISSUE THREAD

**Issue by AntonNiklasson** [eslint/eslint#8146]
I wasn't sure if this already exists or not. Sorry for the noise if this already exists somewhere.

**Please describe what the rule should do:**

Make users aware about a potential overwrite when the spread operator is involved in creating a new 
object.

**What category of rule is this?**

[  ] Enforces code style
[X] Warns about a potential error
[  ] Suggests an alternate way of doing something
[  ] Other (please specify:)

**Code examples that this rule will warn about:**

```js
const foo = { prop: 1 };
const bar = { prop: 2, ...foo };

// bar = { prop: 1 }
```

I often use the spread operator in my redux reducers like this:

```
// foo is set to 'bar'
const reducer = (state, action) => {
  return {
    ...state,
    foo: 'bar'
  }
}

// This would probably cause some unwanted behaviour:
const reducer = (state, action) => {
  return {
    foo: 'bar',
    ...state
  }
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a core feature of the language, and not really specific to any kind of framework or plugin as far as I know.

**Comment by platinumazure** [comment]
Just to make sure, is the goal to ensure that object spreads come first when declaring a new object literal?

**Comment by not-an-aardvark** [comment]
Wouldn't this be a useful way to declare defaults for an object?

```js


const resolvedOptions = { 
  foo: 1, // foo defaults to 1, but a user-provided value takes precedence.
  ...optionsFromUser
};
```

**Comment by AntonNiklasson** [comment]
@platinumazure Yes, that is the intended behaviour.

@not-an-aardvark That is definitely an awesome way of using it! I wasn't sure about these two, I opened this mostly for discussing it. It would probably use it as a warning, but maybe that's stupid.

**Comment by platinumazure** [comment]
@AntonNiklasson It does seem that users might have a legitimate use case for both kinds (spread-first to allow API overrides, or spread-last to allow user overrides). So a rule that consistently enforces one or the other might not always be productive. Depending on usage patterns, a user might have to disable as many as half of the reports of this rule.

Between that and the fact that object spread/rest is still stage 3, I guess I'm not sure I see the utility of adding this rule to ESLint core at this point. So I guess I'm weakly :-1: to this but could certainly be persuaded to support this if a good way to identify when spread-first vs spread-last should be enforced.

**Comment by not-an-aardvark** [comment]
I agree, I don't think this is a good thing to report. Maybe we can reconsider after object rest/spread reaches stage 4 and idiomatic patterns develop more.

Regardless, thanks for the suggestion!

**Comment by platinumazure** [comment]
@AntonNiklasson If we don't accept this into core, please feel free to implement in a plugin or as a custom rule if you find this useful for your own work, and please stop by our [Gitter chat](https://gitter.im/eslint/eslint) if you need help. Thanks for the suggestion!

**Comment by AntonNiklasson** [comment]
Thanks for the feedback guys! I agree with all of that, good points. 👍
